### ISSUE TITLE
Feature Request: obtain temporary access credential for Kubernetes clusters in terraform

### ISSUE THREAD

**Issue by shankerwangmiao** [aliyun/terraform-provider-alicloud#5383]
Hi there,

It is safer to use short-term credential to access Kubernetes clusters on Aliyun. Currently, the access credential can only be obtained via the [`certificate_authority`](https://registry.terraform.io/#certificate_authority) attribute of the resources `alicloud_cs_*_kubernetes`, which is valid for 3 years, and is relatively not safe for automation use. The API [`DescribeClusterUserKubeconfig`](https://help.aliyun.com/document_detail/183915.html) can generate credentials which are valid only for less than 3 days and thus much safer.

It is suggested that a new data source can be added, which calls this API and gives out a temporary credential.

**Comment by Pangjiping** [comment]
Do you want it to expire automatically or update the latest kubeconfig every time you refresh？

**Comment by shankerwangmiao** [comment]
> Do you want it to expire automatically or update the latest kubeconfig every time you refresh？

I believe that the data source can simply expose the latest kubeconfig as its attribute each time it is refreshed, and let users to deal with the content.

The way I would like to use it is that parse the kubeconfig in terraform using `yamldecode` and feed its content to the kubernetes provider, so that other kubernetes operations are performed using this short time credential to improve the security.

It would be better if the kubeconfig can also be parsed in the provider and an attribute like [`certificate_authority`](https://registry.terraform.io/providers/aliyun/alicloud/latest/docs/resources/cs_managed_kubernetes#certificate_authority) can be exposed.

**Comment by Pangjiping** [comment]
> > Do you want it to expire automatically or update the latest kubeconfig every time you refresh？
> 
> I believe that the data source can simply expose the latest kubeconfig as its attribute each time it is refreshed, and let users to deal with the content.
> 
> The way I would like to use it is that parse the kubeconfig in terraform using `yamldecode` and feed its content to the kubernetes provider, so that other kubernetes operations are performed using this short time credential to improve the security.
> 
> It would be better if the kubeconfig can also be parsed in the provider and an attribute like [`certificate_authority`](https://registry.terraform.io/providers/aliyun/alicloud/latest/docs/resources/cs_managed_kubernetes#certificate_authority) can be exposed.

I understand that you want the kubeconfig information of the cluster to be exposed in the state file instead of being written to a local file. 

we will evaluate this requirement and implement it as soon as possible.

**Comment by shankerwangmiao** [comment]
That is correct. I don't think there is any need for alicloud provider to write any local files to the host that is executing terraform operations. If the user would like to do so, then [local_file](https://registry.terraform.io/providers/hashicorp/local/latest/docs/resources/file) and [local_sensitive_file](https://registry.terraform.io/providers/hashicorp/local/latest/docs/resources/sensitive_file) can be utilized.

**Comment by Pangjiping** [comment]
The kubeconfig with expiration time will be supported in this pr #5397 .

So far we are still discussing how to better expose the kubeconfig in the state file. Due to some historical reasons, this field may not be supported in the short term. You could manage cluster's kubeconfig with field `kube_config` in resource and field `kube_config_file_prefix` in datasource right now.

**Comment by shankerwangmiao** [comment]
Hi,

Let me put it more clearly. I believe it would be better to add a new data source instead of embedding this function inside of kubernetes cluster datasources as an attribute.

**Comment by shankerwangmiao** [comment]
[aws_eks_cluster_auth](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/eks_cluster_auth) would be an example, where the temporary authentication credential is fetched via a separate data source.

**Comment by Pangjiping** [comment]
> [aws_eks_cluster_auth](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/eks_cluster_auth) would be an example, where the temporary authentication credential is fetched via a separate data source.

Hi here #5404

**Comment by shankerwangmiao** [comment]
> > [aws_eks_cluster_auth](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/eks_cluster_auth) would be an example, where the temporary authentication credential is fetched via a separate data source.
> 
> Hi here #5404

Hi, many thanks for your work. I have some comments on the design of the module.

First, I'm a bit confused about the design of the arguments. Why `ids` and `name_regex` are introduced? I don't think it is necessary to provide this function. Just fetching a temporary credential for a certain cluster by its id is enough. Data sources and resources are fundamental building blocks of a terraform infrastructure configuration and should do **one just enough thing**. I suggest the data source can be used as follows:

``` terraform
data "alicloud_cs_cluster_auth" "credentials" {
  cluster_id = "abcde....<cluster id goes here>"
  temporary_duration_minutes = 60
 }
```

If a user would like to fetch credentials for more than one clusters, the following can be used and there's no need for the `alicloud_cs_cluster_auth` data source to provide the function to list and filter the clusters:

```terraform
data "alicloud_cs_cluster_auth" "credentials" {
  for_each = toset(["id1", "id2"])
  cluster_id = each.key
  temporary_duration_minutes = 60
 }
```

Or:

```terraform
data "alicloud_cs_managed_kubernetes_clusters" "clusters" {
  name_regex = "<some regex>"
}
data "alicloud_cs_cluster_auth" "credentials" {
  for_each = toset(data.alicloud_cs_managed_kubernetes_clusters.clusters.ids)
  cluster_id = each.value
  temporary_duration_minutes = 60
 }
```

Secondly, the name of the attributes `tokens` and `token` is not appropriate. Since the term `token` has its specific meaning regarding Kubernetes authentication that is the Bearer token used by the client to communicate with the API server, and here what is returned in the `config` field by `DescribeClusterKubeConfigWithExpiration` is not a Bearer token, but a Kubeconfig configuration containing a short term certificate and private key pair, the name `token` should be changed to `kubeconfig`.

Thirdly, since the attribute `token` contains authentication credentials, it should be marked as sensitive.

Thanks again for your work and hope my suggestion can be considered. Wish you a happy Mid-Autumn Festival.

**Comment by Pangjiping** [comment]
> 

you are right. It's available by field `ids`. Multi `ids` and `name_regex` may be used in more scenarios. If you want to expose one cluster's kubeconfig, you could get it by field `ids` which is a argument.

```
data "alicloud_cs_managed_kubernetes_clusters" "clusters" {
  name_regex = "<some regex>"
}
data "alicloud_cs_cluster_auth" "credentials" {
  for_each = toset(data.alicloud_cs_managed_kubernetes_clusters.clusters.ids)
  ids = [each.key]
  temporary_duration_minutes = 60
 }
```

I will optmize schema and doc according your suggestion. Thanks a lot.

 Happy Mid-Autumn Festival.

**Comment by Pangjiping** [comment]
Sorry error in last reply. 

Usage Emaple:
```terraform
data "alicloud_cs_managed_kubernetes_clusters" "clusters" {
  name_regex = "<some regex>"
}
data "alicloud_cs_cluster_auth" "credentials" {
  ids = data.alicloud_cs_managed_kubernetes_clusters.clusters.ids
  temporary_duration_minutes = 60
 }
```

It's not necessary to support multi cluster for once?

**Comment by Pangjiping** [comment]
Hi, I have made overall optimization according to your suggestion and added the usage of linkage with other datasources like `alicloud_cs_managed_kubernetes_clusters`. 

You are correct! This datasource does not need to use regular or array to obtain cluster creds in batches. You could review this PR #5404 if you have any time , this feature will be released as soon as 1.186.0

**Comment by shankerwangmiao** [comment]
Hi, I've proposed some changes to the doc. Many thanks for your work.

**Comment by Pangjiping** [comment]
> Hi, I've proposed some changes to the doc. Many thanks for your work.

It is obviously more reasonable to use a datasource to manage cluster creds, which saves many time about updating schema for each resources. Thank you for your contribution.

Enjoy the last day of the holiday ^_^
